core/app/models/spree/credit_card.rb

```diff
@@ -151,10 +151,10 @@ def require_card_numbers?
     end
 
     def ensure_one_default
-      if self.user_id && self.default
-        CreditCard.where(default: true).where.not(id: self.id).where(user_id: self.user_id).each do |ucc|
-          ucc.update_columns(default: false)
-        end
+      if user_id && default
```

Comment:
Instead of dynamically fixing data on write, why not provide a migration that does it and add some constraints that allow to persist this invariant?

Fixing data on operations that intend to only read is an antipattern as performance over a wide range of users is indeterministic. 

Also you de-facto have a code branch that per user only runs once but you have to maintain it.

What if in future you want to change this code and there is "still" bad data on rest in your DB? Will you maintain this "fix data on first read" semantics for the inifine future?

For this scenario a one time global fix + constraint is the only way to ensure you do not add ever growing friction to the project.